.ci/test_r_package_windows.ps1

```diff
@@ -0,0 +1,107 @@
+# Download a file and retry upon failure. This looks like
+# an infinite loop but CI-level timeouts will kill it
+function Download-File-With-Retries {
+  param(
+    [string]$url,
+    [string]$destfile
+  )
+  do {
+    Write-Output "Downloading '${url}'"
+    sleep 5;
+    (New-Object System.Net.WebClient).DownloadFile($url, $destfile)
+  } while(!$?);
+}
+
+$env:R_LIB_PATH = "C:/RLibrary"
+$env:PATH = "$env:R_LIB_PATH/Rtools/bin;" + "$env:R_LIB_PATH/R/bin/x64;" + "$env:R_LIB_PATH/miktex/texmfs/install/miktex/bin/x64;" + $env:PATH
+$env:CRAN_MIRROR = "https://cloud.r-project.org/"
+
+cd $env:BUILD_SOURCESDIRECTORY
+tzutil /s "GMT Standard Time"
+[Void][System.IO.Directory]::CreateDirectory($env:R_LIB_PATH)
+
+if ($env:COMPILER -eq "MINGW") {
+  Write-Output "Telling R to use MinGW"
+  $install_libs = "$env:BUILD_SOURCESDIRECTORY\R-package\src\install.libs.R"
+  ((Get-Content -path $install_libs -Raw) -replace 'use_mingw <- FALSE','use_mingw <- TRUE') | Set-Content -Path $install_libs
+}
+
+# set up R if it doesn't exist yet
+if (!(Get-Command R.exe -errorAction SilentlyContinue)) {
+
+    Write-Output "Downloading R and Rtools"
+
+    # download R and RTools
+    Download-File-With-Retries -url "https://cloud.r-project.org/bin/windows/base/old/$env:R_WINDOWS_VERSION/R-$env:R_WINDOWS_VERSION-win.exe" -destfile "R-win.exe"
+    Download-File-With-Retries -url "https://cloud.r-project.org/bin/windows/Rtools/Rtools35.exe" -destfile "Rtools.exe"
+
+    # Install R
+    Write-Output "Installing R"
+    Start-Process -FilePath R-win.exe -NoNewWindow -Wait -ArgumentList "/VERYSILENT /DIR=$env:R_LIB_PATH/R /COMPONENTS=main,x64" ; Check-Output $?
+    Write-Output "Done installing R"
+
+    Write-Output "Installing Rtools"
+    Start-Process -FilePath Rtools.exe -NoNewWindow -Wait -ArgumentList "/VERYSILENT /DIR=$env:R_LIB_PATH/Rtools" ; Check-Output $?
+    Write-Output "Done installing Rtools"
+
+}
+
+# MiKTeX and pandoc can be skipped on Azure builds, since we don't
+# build the package documentation there
+if ($env:AZURE -ne "true") {
```

Comment:
I actually literally meant `AZURE` here. I am not sure yet if the issue is because of Azure or `MSVC` but I really expect that it is  Azure.

I can't find the old comment, so here is the summary...when I was testing earlier commits here, the Azure builds for the R package were  taking 20-25  minutes, even after #2957 . So I decided to change the `R CMD check` commands for AZURE builds to not check any docs:

```
if ($env:AZURE -eq "true") {
  Write-Output "Running R CMD check without checking documentation"
  R.exe CMD check --no-multiarch --no-examples --no-manual --ignore-vignettes ${PKG_FILE_NAME} ; Check-Output $?
```

Once I did that, I was also able to save time by cutting out MiKTeX and pandoc installation. If you are running with `--no-examples --no-manual --ignore-vignettes`, you never need to create the package documentation.

I am ok with this because the docs are already checked in Linux / Mac builds and on the AppVeyor build.

Another possibility that the jobs took SO LONG to run might be based on how Azure does virtualization vs how AppVeyor does it.  I have seen that we use `/MP` [for MSVC builds](https://github.com/microsoft/LightGBM/blob/master/CMakeLists.txt#L196).

From the MSVC docs for `/MP` ([docs](https://docs.microsoft.com/en-us/cpp/build/reference/mp-build-with-multiple-processes?view=vs-2019)):

> /MP[processMax]
> If you omit the processMax argument, the compiler retrieves the number of effective processors on your computer from the operating system, and creates a process for each processor.

If you'd like, I can move any `if ($env:AZURE...` stuff out of this and into #2965 if you'd prefer. I originally didn't do that because I thought it would be complicated to remember to add it back in the right places.

Reply:
OK, I see. Thanks for the explanation! But what about MinGW on Azure? Does it take huge amount of time too?

Reply:
I never tried MinGW on Azure. Let me add a build and see.

Reply:
seems we still have issues with CTAN mirror connectivity from Azure

![image](https://user-images.githubusercontent.com/7608904/79415209-9dfa1d80-7fa4-11ea-929e-8bb72cabf010.png)

Reply:
The problem is that the URL Azure trying to access is `http://ftp.yzu.edu.tw/CTAN/systems/win32/miktex/tm/packages/miktex-dvisvgm-bin-2.9.tar.lzma%22`. Note `%22` at the end.

And `https://miktex.org/download/win/miktexsetup-x64.zip'` - quote sign at the end.

Reply:
Hmm, maybe we can try to pin CRAN mirror?

> --remote-package-repository=url
Download from the specified URL. Use --list-repositories to download an up-to-date list of possible repositories.
https://docs.miktex.org/manual/miktexsetup.html

https://github.com/MiKTeX/miktex/issues/495

Reply:
gaaaah I was looking for this! Thank you.

I chose one from University of Illinois because it seems to have good uptime recently ([CTAN mirror status](http://dante.ctan.org/mirmon/)), although that site is pretty limited so 🤷 

```
$env:CTAN_MIRROR = "https://ctan.math.illinois.edu/systems/win32/miktex/tm/packages/"
```

Reply:
Wow, it is working now! Thanks a lot for the fix!

Reply:
> Haha, hello `chocolatey`!
> 
> ```
> -- The C compiler identification is GNU 8.1.0
> -- The CXX compiler identification is GNU 8.1.0
> -- Check for working C compiler: C:/ProgramData/chocolatey/bin/gcc.exe
> -- Check for working C compiler: C:/ProgramData/chocolatey/bin/gcc.exe - works
> ```

haha NOOOOOOOOOOO

I probably need to put the RTools-bundled MinGW on `PATH`

Reply:
haha yes I kind of agree. I think pinning to a CTAN mirror will still be useful, it's one less thing that can change in our builds.

Reply:
changed back to "don't run R docs on Azure" here: https://github.com/microsoft/LightGBM/commit/d843795087e363ffe7bbc53c502fdbe8d36df9fc